DESIGN.md

```diff
@@ -0,0 +1,370 @@
+            }
+        }
+    }
+}
+```
+
+Nested interfaces are named based on the response name, not the field type or field name. Response names are guaranteed to be unique within a selection set, and it also seems to make more semantic sense. (A field `author` would refer to `Author` for example, and not to `User`.)
+
+Aliases map naturally onto this model as well, so there is no need to use further qualifications I think (like `R2Character`):
+
+```graphql
+query TwoHeroes {
+  r2: hero {
+    name
+  }
+  luke: hero(episode: EMPIRE) {
+    name
+  }
+}
+```
+
+```java
+class TwoHeroesQuery implements GraphQLQuery<TwoHeroesQuery.Data> {
+    interface Data {
+        R2 r2();
+        interface R2 {
+            String name();
+        }
+        Luke luke();
+        interface Luke {
+            String name();
+        }
+    }
+}
+```
+
+I don't think we ever want `r2` and `luke` to refer to the same type here. So I don't think we need the dedupe behavior from [PR#4](https://github.com/apollostack/apollo-android/pull/4). Even if they select the same fields now, that is accidental from the perspective of the type system, so we still want the types to be separate to avoid relying on these accidental shared fields.
+
+In GraphQL, fragments offer an explicit way to share selections.
+
+## Fragments
+
+```graphql
+query TwoHeroes {
+  r2: hero {
+    ...HeroDetails
+  }
+  luke: hero(episode: EMPIRE) {
+    ...HeroDetails
+  }
+}
+
+fragment HeroDetails on Character {
+  name
+}
+```
+
+This would generate an interface `HeroDetails` that could be reused between selections and between operations. This is also important because it allows you to write code that isn't tied to a specific query, like a UI component that knows how to display any hero's details.
+
+The simplest mapping may seem to be to let `R2` and `Luke` extend `HeroDetails`:
+
+```java
+class TwoHeroesQuery implements GraphQLQuery<TwoHeroesQuery.Data> {
+    interface Data {
+        R2 r2();
+        interface R2 extends HeroDetails {}
+        Luke luke();
+        interface Luke extends HeroDetails {}
+    }
+}
+
+interface HeroDetails {
+    String name();
+}
+```
+
+But this model gets a lot more complicated when nested selection sets are taken into account:
+
+```graphql
+query TwoHeroes {
+  r2: hero {
+    ...HeroDetails
+    friends {
+      appearsIn
+    }
+  }
+  luke: hero(episode: EMPIRE) {
+    ...HeroDetails
+    friends {
+      id
+    }
+  }
+}
+
+fragment HeroDetails on Character {
+  name
+  friends {
+    name
+  }
+}
+```
+
+This would have to generate something like:
+
+```java
+interface R2 extends HeroDetails {
+  String name();
+  List<Friend> friends();
+  interface Friend extends HeroDetails.Friend {
+    List<Episode> appearsIn();
+  }
+}
+
+interface Luke extends HeroDetails {
+  String name();
+  List<Friend> friends();
+  interface Friend extends HeroDetails.Friend {
+    String id();
+  }
+}
+
+interface HeroDetails {
```

Comment:
Yes, all `.graphql` files are processed together, and GraphQL validation rules will make sure a fragment has only been defined once (and generate a validation error otherwise).